python/glow/wgr/linear_model/functions.py

```diff
@@ -112,6 +116,46 @@ def assemble_block(n_rows: Int, n_cols: Int, pdf: pd.DataFrame,
         return X
 
 
+@typechecked
+def get_irls_pieces(X: NDArray[Float], y: NDArray[Float], alpha: Float,
+                    p0: Float) -> (NDArray[Float], NDArray[Float], NDArray[Float]):
+    """
+    Fits a logistic regression model logit(p(y|X)) ~ X*beta with L2 shrinkage C = 1/alpha, using scikit-learn.  Returns
+    the vector beta, the matrix transpose(X)*diag(p(1-p))*X, and the vector transpose(X)*(y - p), which are the pieces
+    needed to perform one step of the IRLS algorithm.
+
+    Args:
+         X: Matrix of n samples m features.  Matrix X is expected to include a constant intercept = 1 term as the first
+          column.
+         y : Binary response vector of  length n
+         alpha: Shrinkage parameter to be used in the fit
+         p0: observed population prevalence of label y.
+
+    Returns:
+        beta : m length array representing coefficients found from the fit, including the intercept term as the first
+        element.
+        XtGX : m by m matrix representing transpose(X)*diag(p(1-p))*X.
+        XtY : m length array representing transpose(X)*(y - p)
+
+    """
+    if y.sum() > 0:
+        clf = sklearn.linear_model.LogisticRegression(C=1 / alpha,
+                                                      solver='lbfgs',
+                                                      fit_intercept=True,
```

Comment:
If we just include the intercept in X, then the sklearn model will shrink the intercept term, which we don't want, but if we tell it to fit the intercept it will do so without shrinkage, and we can pull out the estimate for the intercept separately from the coefficients.  It's a little awkward and I think we should include more explicit handling of the intercept term on the covariates in general, let's discuss on our call.

Reply:
Is there any reason that a user would not want to include an intercept? It seems wrong to me that we both require users to include it and unconditionally include it ourselves. Maybe it would be simpler to instruct users to not include an intercept. That would also match the expectation for quantitative traits, where we expect that the covariates matrix does not include an intercept (albeit for different reasons).

Reply:
Sorry, going back and reading my comment above I realize it is not very clear.  We are expecting that the covariates DataFrame has an intercept column, and we are now enforcing that with the input validation function so that things like assigning the `sort_key` for the intercept term just work as normal elsewhere in the code, but we are not including that intercept term in the call to sklearn LogisticRegression in`get_irls_pieces` because we don't want to apply shrinkage to the intercept.  Instead, we let sklearn fit an intercept separately with `fit_intercept=True` and then take the parameter estimate from the `clf.intercept_` and insert it into the appropriate spot in the `beta` array.  I like your suggestion that we should instead instruct the user to not include an intercept term and add it for them, so that we can also ensure that it is the first column in the covariates DataFrame and that it is called 'intercept'.  This will also more closely match the use pattern for `RidgeRegression`, like you said.